src/Common/src/System/Net/Security/NegotiateStreamPal.Unix.cs

```diff
@@ -233,11 +233,11 @@ internal static string QueryContextAuthenticationPackage(SafeDeleteContext secur
         }
 
         internal static SecurityStatusPal InitializeSecurityContext(
-            SafeFreeCredentials credentialsHandle,
+            ref SafeFreeCredentials credentialsHandle,
             ref SafeDeleteContext securityContext,
             string spn,
             ContextFlagsPal requestedContextFlags,
-            SecurityBuffer[] inSecurityBufferArray,
+            Span<SecurityBuffer> inSecurityBufferArray,
```

Comment:
This may be beyond the scope of this change, but I'll mention it anyway...

I'm not sure it really makes sense to expose SecurityBuffer as a concept in the PAL layers, e.g. NegotiateStreamPal and SslStreamPal.

As I understand it, this all comes from the SCHANNEL APIs originally. The whole mechanism is very general purpose.

But if you look at the usage of this particular API (NegotiateStreamPal.InitializeSecurityContent), it seems like it's always used in a very specific way:

(1) One security buffer for the incoming blob, if present
(2) One security buffer for the channel binding, if present

And in fact, the Unix implementation of the Pal assumes exactly this and doesn't support anything else.

So why doesn't this API simply take arguments for (a) incoming blob and (b) ChannelBinding, and not expose SecurityBuffer here at all?

I suspect similar things could and should be done across NegotiateStreamPal and SslStreamPal, though I haven't looked at the specific usages. 

In short, let's kill SecurityBuffer entirely and just use the platform-specific APIs and structures within the PAL layers as appropriate.

Reply:
Or perhaps a more realistic first step: Remove SecurityBuffer usage from the PAL APIs and make it Windows-specific. Later, we can modify the Windows-specific implementation to remove this or make it a struct or whatever makes sense.

Looking at the usage, there actually aren't that many uses of SecurityBuffer outside of Windows-specific code today.

Reply:
I'm not entirely following.  Many of the places SecurityBuffers are constructed is in shared code, namely SecureChannel.cs and NTAuthentication.Common.cs, which are used on Unix despite the names.

I'm going to leave this as-is for now; seems like a more targeted PAL cleanup.  I can open an issue if you like.

Reply:
> Many of the places SecurityBuffers are constructed is in shared code, namely SecureChannel.cs and NTAuthentication.Common.cs, which are used on Unix

Yes, but it's actually just two functions that use SecurityBuffers -- SecureChannel.GenerateToken and NTAuthentication.GetOutgoingBlob. And they only call a couple PAL functions -- InitializeSecurityContent, AcceptSecurityContext, and CompleteAuthToken (which BTW does nothing on Unix, and so I'm not sure why it's even exposed in the PAL, but that's a whole other issue).

So my suggestion is to change these PAL functions to not use SecurityBuffers, instead just pass appropriate args (incoming blob, ChannelBinding, etc) and construct the SecurityBuffers themselves. Then we can make SecurityBuffer itself Windows-only. This avoids SecurityBuffer allocation entirely on Unix, and is a step toward reworking it however we want on Windows.

Reply:
I see.

Ok, done.  I made SecurityBuffer be Windows-only, removing it from the PAL, and I also made it a struct, which eliminated a huge portion of the allocation.